File: solutions/080.rb
#  <p>It is well known that if the square root of a natural number is not an intege
# r, then it is irrational. The decimal expansion of such square roots is infinite
#  without any repeating pattern at all.</p>
# <p>The square root of two is $1.41421356237309504880\\cdots$, and the digital sum
#  of the first one hundred decimal digits is $475$.</p>
# <p>For the first one hundred natural numbers, find the total of the digital sums
#  of the first one hundred decimal digits for all the irrational square roots.</p
# >

# Solution for Project Euler Problem 80
require 'bigdecimal'

# Initialize total sum
total_digital_sum = 0
NUM_DECIMAL_DIGITS = 100

# Set precision: 1 (integer) + 100 (decimals) + 19 guard digits
CALCULATION_PRECISION = NUM_DECIMAL_DIGITS + 20 # 120 significant digits
TEN_POW_NUM_DECIMAL_DIGITS = BigDecimal('10').power(NUM_DECIMAL_DIGITS)

(1..100).each do |num|
  # Check if √num is an integer (perfect square)
  root_test = Math.sqrt(num)
  next if root_test == root_test.floor

  # Compute square root with high precision
  sqrt_val = BigDecimal(num).sqrt(CALCULATION_PRECISION)

  # Scale by 10^100 and take floor
  scaled_sqrt = (sqrt_val * TEN_POW_NUM_DECIMAL_DIGITS).floor
  scaled_str = scaled_sqrt.to_i.to_s # Convert to string

  # Since √n < 10, scaled_sqrt has 100 or 101 digits
  # If 101 digits, first digit is integer part; take last 100
  decimal_digits = scaled_str.length > NUM_DECIMAL_DIGITS ? scaled_str[-NUM_DECIMAL_DIGITS..-1] : scaled_str.rjust(NUM_DECIMAL_DIGITS, '0')

  # Sum the digits
  current_sum = decimal_digits.chars.map(&:to_i).sum
  total_digital_sum += current_sum

  # Verification for √2
  puts "√#{num} digit sum: #{current_sum}" if num == 2
end

puts "Total sum: #{total_digital_sum}"

File: python/080.py
#!/usr/bin/env python3
"""
Square root digital expansion (Problem 80)

Compute the total of the digital sums of the first one hundred digits of the
square roots of the first one hundred natural numbers, excluding perfect
squares. Uses integer arithmetic to avoid precision loss.
"""

DIGITS = 100
SCALE = 10 ** DIGITS

def integer_sqrt(n):
    """Integer square root using Newton's method."""
    if n == 0:
        return 0
    x = n
    y = (x + 1) // 2
    while y < x:
        x = y
        y = (x + n // x) // 2
    return x

def digital_sum_for_sqrt(n):
    """Calculate digital sum of first DIGITS of sqrt(n), excluding perfect squares."""
    root = integer_sqrt(n)
    if root * root == n:  # skip perfect squares
        return 0

    scaled_root = integer_sqrt(n * SCALE * SCALE)

    integer_part = scaled_root // SCALE
    fractional_part = scaled_root % SCALE

    digits = str(integer_part) + str(fractional_part).rjust(DIGITS, '0')
    total = 0
    for ch in digits[:DIGITS]:
        total += ord(ch) - 48  # Convert char to digit
    return total

total = 0
for n in range(1, 101):
    total += digital_sum_for_sqrt(n)

print(total)
